File: Array/Python/reverse.py
"""
Time Complexity: O(n) where n is the length of the input string.
This is because the slicing operator [::-1] iterates through the entire string once to create a new reversed string.

Space Complexity: O(n) where n is the length of the input string.
This is because a new string is created to store the reversed string.
"""
# Function to reverse a string
def reverse(string):
    return string[::-1]

print(reverse("Hey how you doing?"))

File: Array/C++/reverse.cpp
#include <iostream>
#include <algorithm>

using namespace std;

/*
Time Complexity: O(n/2) where n is the length of the input string.
This is because the std::reverse function iterates through half of the string, swapping the elements at the beginning and end, and moving towards the middle.

Space Complexity: O(1) as the method is modifying the original string passed as parameter and no new memory is allocated.
*/
std::string reverse(std::string str) {
    std::reverse(str.begin(), str.end());
    return str;
}

int main() {

    std::string rstring = reverse("Hey, how you doing");
    std::cout << rstring << std::endl;
}